pkg/slo-controller/nodeslo/extender_plugin.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeslo
+
+import (
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
+)
+
+var (
+	globalNodeSLOMergedExtender = map[string]NodeSLOMergedPlugin{}
+)
+
+func RegisterNodeSLOMergedExtender(name string, extender NodeSLOMergedPlugin) error {
+	if _, exist := globalNodeSLOMergedExtender[name]; exist {
+		return fmt.Errorf("node merged plugin %s already exist", name)
+	}
+	globalNodeSLOMergedExtender[name] = extender
+	return nil
+}
+
+func calculateExtensionsCfgMerged(oldCfg extension.ExtensionCfg, configMap *corev1.ConfigMap, recorder record.EventRecorder) extension.ExtensionCfg {
+	oldCfgCopy := *oldCfg.DeepCopy()
+	if oldCfgCopy.ClusterStrategy == nil {
+		oldCfgCopy.ClusterStrategy = &slov1alpha1.ExtensionsMap{}
+	}
+	if oldCfgCopy.ClusterStrategy.Object == nil {
+		oldCfgCopy.ClusterStrategy.Object = make(map[string]interface{})
+	}
+	mergedCfg := *oldCfgCopy.DeepCopy()
+	for name, extender := range globalNodeSLOMergedExtender {
+		newCfg, err := extender.MergeNodeSLO(mergedCfg, configMap, recorder)
+		if err != nil {
+			klog.Warningf("run merge nodeSLO extender %v failed, error %v", name, err)
+			continue
+		}
+		if !reflect.DeepEqual(newCfg, mergedCfg) {
+			mergedCfg = *newCfg.DeepCopy()
+		}
+		klog.V(5).Infof("run merge nodeSLO extender %v success, extensionCfg detail update %v", name, mergedCfg)
+	}
+	if reflect.DeepEqual(oldCfgCopy, mergedCfg) {
+		return oldCfg
+	}
+	return mergedCfg
+
+}
+
+func getExtensionsConfigSpec(node *corev1.Node, cfg *extension.ExtensionCfg) *slov1alpha1.ExtensionsMap {
+	extMap := slov1alpha1.ExtensionsMap{}
+	if cfg == nil || cfg.ClusterStrategy == nil || cfg.ClusterStrategy.Object == nil {
+		return &extMap
+	}
+	for name, extender := range globalNodeSLOMergedExtender {
+		extKey, extStrategy, err := extender.GetNodeSLO(node, cfg)
```

Comment:
here we can check whether to use clusterStrategy or nodeStrategy.
so that we don't need to implement in each extender plugin.

Reply:
I guess it's better implemented in each GetNodeSLOExtension, as only after parsing the cfg can we know whether an extension has clusterStrategy  or nodeStrategy. And parsing the cfg need extra info and cannot done in precheck.